src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -141,11 +141,15 @@ private bool TryRewriteTagHelper(Block tagBlock, RewritingContext context)
             {
                 // We're in an end tag helper block.
 
-                var tagNameScope = _tagStack.Count > 0 ? _tagStack.Peek().TagName : string.Empty;
+                var parentTagHelper = _tagStack.Count > 0 ? _tagStack.Peek() : null;
 
                 // Validate that our end tag helper matches the currently scoped tag helper, if not we
                 // need to error.
-                if (tagNameScope.Equals(tagName, StringComparison.OrdinalIgnoreCase))
+                if (parentTagHelper != null &&
+                    string.Equals(
+                        parentTagHelper.Prefix + parentTagHelper.TagName,
```

Comment:
Have to based on our convo here: https://github.com/aspnet/Razor/pull/313#discussion_r25537282